### PR TITLE
Fix a memory leak in `DownloadTaskManager` and `DataTaskManager`

### PR DESCRIPTION
**Pull request by ahoppen**
`DownloadTaskManager` had a strong reference to `URLSession`, which retained its delegate, which was the `DownloadTaskManager`. This formed a retain cycle.

Make the reference from `URLSession` to the delegate weak and only keep `DownloadTaskManager` alive as long as `DownloadTask`s need it. This causes the `DownloadTaskManager` to be deallocated once nobody holds a reference to it anymore and all the tasks it manages are finished.

Finally, we need to call `invalidate` on the `URLSession` to tell it that we’re done and that it should release its delegate (which is now the `WeakDownloadTaskManager`wrapper).

This retain cycle was causing a leak in sourcekit-lsp. I verified that the leak no longer exists with this patch.

rdar://125012584

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Sources/Basics/HTTPClient/URLSessionHTTPClient.swift [hunk 1]
```diff
@@ -87,7 +87,35 @@ final class URLSessionHTTPClient {
...
+    ) {
+        dataTaskManager?.urlSession(session, dataTask: dataTask, didReceive: response, completionHandler: completionHandler)
+    }
+
+    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
+        dataTaskManager?.urlSession(session, dataTask: dataTask, didReceive: data)
+    }
+}
+
+private class DataTaskManager: NSObject {
```
Comment (MaxDesiatov) [inline Sources/Basics/HTTPClient/URLSessionHTTPClient.swift:118]: Does this still need to inherit from `NSObject` if it no longer conforms to `URLSessionDataDelegate`?

File: Sources/Basics/HTTPClient/URLSessionHTTPClient.swift [hunk 2]
```diff
@@ -202,18 +240,49 @@ private class DataTaskManager: NSObject, URLSessionDataDelegate {
...
+            session,
+            downloadTask: downloadTask,
+            didWriteData: bytesWritten,
+            totalBytesWritten: totalBytesWritten,
+            totalBytesExpectedToWrite: totalBytesExpectedToWrite
+        )
+    }
+}
+
+private class DownloadTaskManager: NSObject {
```
Comment (MaxDesiatov) [inline Sources/Basics/HTTPClient/URLSessionHTTPClient.swift:285]: Ditto, I hope `NSObject` inheritance is redundant here.

File: Sources/Basics/HTTPClient/URLSessionHTTPClient.swift [hunk 3]
```diff
@@ -223,7 +292,11 @@ private class DownloadTaskManager: NSObject, URLSessionDownloadDelegate {
         self.delegateQueue.name = "org.swift.swiftpm.urlsession-http-client-download-delegate"
         self.delegateQueue.maxConcurrentOperationCount = 1
         super.init()
-        self.session = URLSession(configuration: configuration, delegate: self, delegateQueue: self.delegateQueue)
+        self.session = URLSession(configuration: configuration, delegate: WeakDownloadTaskManager(self), delegateQueue: self.delegateQueue)
+    }
+
+    deinit {
+      session.finishTasksAndInvalidate()
```
Comment (MaxDesiatov) [inline Sources/Basics/HTTPClient/URLSessionHTTPClient.swift:299]: Formatting seems inconsistent here with the rest of the file.

File: Sources/Basics/HTTPClient/URLSessionHTTPClient.swift (hunk 1 above)
Comment (ahoppen) [reply Sources/Basics/HTTPClient/URLSessionHTTPClient.swift:118]: Oh, nice catch 👍🏽

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

File: Sources/Basics/HTTPClient/URLSessionHTTPClient.swift [hunk 4]
```diff
@@ -202,18 +254,61 @@ private class DataTaskManager: NSObject, URLSessionDataDelegate {
...
+
+    func urlSession(
+        _ session: URLSession,
+        downloadTask: URLSessionDownloadTask,
+        didFinishDownloadingTo location: URL
+    ) {
+        downloadTaskManager?.urlSession(session, downloadTask: downloadTask, didFinishDownloadingTo: location)
+    }
+
+    public func urlSession(
```
Comment (MaxDesiatov) [inline Sources/Basics/HTTPClient/URLSessionHTTPClient.swift:302]: `public` function in a `private` type

File: Sources/Basics/HTTPClient/URLSessionHTTPClient.swift [hunk 5]
```diff
@@ -87,7 +87,50 @@ final class URLSessionHTTPClient {
...
+        _ session: URLSession,
+        task: URLSessionTask,
+        willPerformHTTPRedirection response: HTTPURLResponse,
+        newRequest request: URLRequest,
+        completionHandler: @escaping (URLRequest?) -> Void
+    ) {
+        dataTaskManager?.urlSession(session, task: task, willPerformHTTPRedirection: response, newRequest: request, completionHandler: completionHandler)
```
Comment (MaxDesiatov) [inline Sources/Basics/HTTPClient/URLSessionHTTPClient.swift:129]: We need to keep line length under 120 columns per the `.swiftformat` config.
```suggestion
        dataTaskManager?.urlSession(session, task: task, willPerformHTTPRedirection: response, newRequest: request, completionHandler: completionHandler)
```

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Review by MaxDesiatov — APPROVED** [review]
Thanks!

File: Sources/Basics/HTTPClient/URLSessionHTTPClient.swift (hunk 4 above)
Comment (bnbarham) [reply Sources/Basics/HTTPClient/URLSessionHTTPClient.swift]: FWIW I see nothing wrong with `public` visibility inside a `private` type, ie. visibility inside the type is describing the intent of the API and then visibility on the type for how accessible you want it to be. It makes moving (and re-using) that type later much simpler. One argument against that is it's quite likely a private type hasn't had much thought to its API, which could be true 🤷.

**Pull request merged**